src/capmap.cc

```diff
@@ -232,4 +238,38 @@ bool LoadMap::try_combine(void* __capability cap) {
   return false;
 }
 
+bool BasicMap::try_combine(void* __capability cap) {
+  if (!cheri_tag_get(cap) || cheri_is_sealed(cap)) return false;
+  if ((cheri_perms_get(cap) & perms_) != perms_) return false;  // match ALL perms
+
+  ranges_.combine(Range::from_cap(cap));
+  return true;
+}
+
+bool BranchMap::try_combine(void* __capability cap) {
+  if (!cheri_tag_get(cap) || cheri_type_get(cap) > CHERI_OTYPE_SENTRY) return false;
+  if (!(cheri_perms_get(cap) & CHERI_PERM_EXECUTE)) return false;
+
+  // If it's a sentry, we treat it as length 1.
+  if (cheri_is_sentry(cap))
+    ranges_.combine(Range::from_base_length(cheri_address_get(cap), 1));
+  else
+    ranges_.combine(Range::from_cap(cap));
+  return true;
+}
+
+bool PoisonMap::try_combine(void* __capability cap) {
+  if (!cheri_tag_get(cap) || cheri_is_sealed(cap)) return false;
+  if (!(cheri_perms_get(cap) & perms_)) return false;  // match ANY perms
+  if (!poison_.overlaps(Range::from_cap(cap))) return false;
+
+  if (!callback_ || (*callback_)(cap)) {
```

Comment:
Just a suggestion. 
I would add `void * user_data` to the callback, in case some context needs to be passed in.